src/services/hsm.service.js

```diff
@@ -6,16 +6,210 @@
+ */
+
+class HsmSession {
+  /**
+   * @param {hsm.interfaces.Hsm} conn
+   * @param {string} rootKeyname
+   */
+  constructor(conn, rootKeyname) {
+    this.conn = conn;
+    this.rootKey = rootKeyname;
+    /**
+     * @type {Record<number, string>}
+     */
+    this.addressKeys = {};
+  }
+
+  /**
+   * Get the adddress index keyname from the address index
+   * @param {index} index - bip32 Address index
+   * @returns {Promise<string>}
+   */
+  async getAddressKeyName(index) {
+    if (index in this.addressKeys) {
+      return this.addressKeys[index];
+    }
+
+    const addressKeyName = getBip32Keyname(this.rootKey, 5, { index });
+    const changeKeyName = getChangeKeyname(this.rootKey);
+    try {
+      await this.conn.blockchain.createBip32ChildKeyDerivation(
+        hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+        index,
+        false,
+        true,
+        changeKeyName,
+        addressKeyName,
+      );
+    } catch (err) /* istanbul ignore next */ {
+      // After an error, the HSM lib has a runtime issue that will fail the next call
+      // To avoid this we can wait 500ms before continuing.
+      // This may be fixed in the near future.
+      await delay(500);
+      // (CODE: 5022 | NAME: ERR_OBJ_ALREADY_EXISTS)
+      // This error code means the address key was already derived
+      if (err.errorCode === hsm.constants.ERR_OBJ_ALREADY_EXISTS) {
+        // address key already exists, just return it
+        this.addressKeys[index] = addressKeyName;
+        return addressKeyName;
+      }
+      console.error(err);
+      throw err;
+    }
+
+    this.addressKeys[index] = addressKeyName;
+    return addressKeyName;
+  }
+
+  /**
+   * Sign the given data with the given HSM key
+   * @param {HsmSession} hsmSession
+   * @param {Buffer} dataToSignHash
+   * @param {number} addressIndex
+   * @returns {Promise<Buffer>} DER encoded signature
+   */
+  async signData(dataToSignHash, addressIndex) {
+    // Derive the key to the desired address index
+    const htrKeyName = await this.getAddressKeyName(addressIndex);
+
+    // This api receives the pre-hashed data to sign
+    // We request a non-deterministic ECDSA signature
+    // and it returns the DER encoded signature Buffer
+    const signature = await this.conn.blockchain.sign(
+      hsm.enums.BLOCKCHAIN_SIG_TYPE.SIG_DER_ECDSA,
+      hsm.enums.BLOCKCHAIN_HASH_MODE.SHA256,
+      dataToSignHash,
+      htrKeyName,
+    );
+
+    // Dinamo SDK returns v + DER, where v is a parity byte
+    // https://manual.dinamonetworks.io/nodejs/enums/hsm.enums.BLOCKCHAIN_SIG_TYPE.html#SIG_DER_ECDSA
+    // We can safely ignore the parity byte
+    if (signature[0] !== 0x30) {
+      return signature.slice(1);
+    }
+    return signature;
+  }
+
+  /**
+   * Sign the given transaction and return the signatures.
+   * @param {hathorLib.Transaction} tx
+   * @param {hathorLib.Storage} storage
+   * @returns {Promise<SignatureData[]>}
+   */
+  async signTx(tx, storage) {
```

Comment:
Don't we have a similar method in the lib? This seems a method exactly like any other `signTx` method, just with the custom `getSignature/signData`. If we don't have it, I think we should, this is an important method to have in the lib.

Reply:
This method is the equivalent of the `transactionUtils.getSignaturesForTx`, but instead of deriving the main key and signing it uses the dinamo api to do so.
It also uses the cached address keys to avoid deriving the same address twice during the signature process.